ipasir modification without CI integration

nmanthey opened the pull request:
I split my IPASIR patch series into the functional part and the CI integration, such that the first part can be merged already while the integration into CI can still be figured out lateron.

mgudemann commented on src/Makefile:
```diff
@@ -84,4 +84,41 @@ glucose-download:
+	@rm -Rf ../libzip
+	@mv libzip-1.1.2 ../libzip
+	@rm libzip_1.1.2.orig.tar.gz
+
```
we switched to internal miniz so this download is not necessary anymore

mgudemann commented on src/Makefile:
```diff
@@ -84,4 +84,41 @@ glucose-download:
+ipasir-build: ipasir-download
+	# build libpicaso, and create a libipasir.a in ipasir directory
+	# make sure that the ipasir.h file is located there as well (ships with 2016 package)
+	@echo "Build Picaso 961 from ipasir 2016 package"
```
picosat

kroening commented on src/Makefile:
```diff
@@ -84,4 +84,41 @@ glucose-download:
 	@rm glucose-syrup.tgz
 
-.PHONY: minisat2-download glucose-download
+zlib-download:
```
The zlib/libzip download got removed; this seems to be an artefact from an older version.

kroening commented on src/config.inc:
```diff
@@ -21,10 +21,14 @@ BUILD_ENV = AUTO
 #GLUCOSE = ../../glucose-syrup
 #SMVSAT =
 
+LIBZIPLIB = ../../libzip/lib/.libs/libzip.a ../../zlib/libz.a
```
Not in use any more.

kroening commented on src/common:
```diff
@@ -150,6 +150,64 @@ else
+ifndef MINISAT
+ifndef MINISAT2
+ifndef GLUCOSE
+MINISAT2 = ../../minisat-2.2.1
```
This will be the only path in common -- a bit of a deviation from the current scheme.
I see you want to be able to switch solvers by giving a command line option to make.
Can that be achieved in any other way?

nmanthey replied:
I am not sure this can be achieved in a different way, as you only want to set the default path in case none of the other variables is set already.

tautschnig replied:
Shouldn't we just fail if no solver has been set? No need for defaults, the user should be forced to choose at least one. I'd do
````make
ifeq ($(BOOLEFORCE)$(CHAFF)$(GLUCOSE)$(IPASIR)$(LINGELING)$(MINISAT)$(MINISAT2)$(PICOSAT)$(PRECOSAT),)
  $(error Please select a SAT solver)
endif
````

nmanthey commented:
I'll remove the zip based code, and update the pull request.

nmanthey commented:
I am not aware of a broad discusion whether variable freezing should become part of the interface, as IPASIR is designed to be simple (and hence SAT solvers can be easily integrated into applications without much knowledge about solver internals). I assume this pull request is not the best place to have this discussion.

The reason for adding the interface here is to allow CBMC to link against "recent" SAT solvers easily (also in the future). In case the interface is improved, the integration in CBMC could be patched again.

nmanthey commented:
I'd love to get the three patches in before 8th of May, so that I can use the master branch of the repository to submit CBMC to the SAT competition as part of the incremental SAT track. Otherwise, I'd have to use my fork.

mgudemann commented on src/cbmc/Makefile:
```diff
@@ -37,7 +37,7 @@ OBJ += ../ansi-c/ansi-c$(LIBEXT) \
       ../assembler/assembler$(LIBEXT) \
-      ../solvers/solvers$(LIBEXT) \
+      ../solvers/solvers$(LIBEXT) $(LIBSOLVER) \
       ../util/util$(LIBEXT) \
```
please put this on its own line, analogous for the other makefiles

mgudemann commented on src/solvers/sat/satcheck_ipasir.h:
```diff
@@ -0,0 +1,56 @@
+  satcheck_ipasirt();
+  virtual ~satcheck_ipasirt();
+
+  virtual const std::string solver_text();
```
please mark as `override`

mgudemann commented:
@nmanthey I did not intend to start this discussion, just curious about this. The failed tests are at least for the `--show-symbol-table` cases due to the exit code not matching, maybe a problem with picosat exiting early?
for building I used `env IPASIR=../../ipasir env LIBSOLVER=../../ipasir/libipasir.a make`, `env IPASIR=../../ipasir env LIBSOLVER=$(pwd)/ipasir/libipasir.a make`  did not work

peterschrammel commented on src/common:
```diff
@@ -150,6 +150,64 @@ else
 endif
 
+# select default solver to be minisat2 if no other is specified
+ifndef IPASIR
```
Maybe use `!defined(...) && ...` to avoid the if cascade.

peterschrammel commented on src/solvers/sat/satcheck_ipasir.cpp:
```diff
@@ -0,0 +1,314 @@
+
+/*******************************************************************\
+
+Function: satcheck_minisat_no_simplifiert::solver_text
```
documentation does not match function

peterschrammel commented on src/solvers/sat/satcheck.h:
```diff
@@ -48,6 +77,13 @@ typedef satcheck_minisat1t satcheck_no_simplifiert;
+
+#include "satcheck_ipasir.h"
+
+typedef satcheck_ipasirt satcheckt;
```
This will only work with `--no-sat-preprocessor` (or `--stop-on-fail` in most cases).

peterschrammel commented on src/Makefile:
```diff
@@ -84,4 +84,17 @@ glucose-download:
+	# build libpicosat, and create a libipasir.a in ipasir directory
+	# make sure that the ipasir.h file is located there as well (ships with 2016 package)
+	@echo "Build Picosat 961 from ipasir 2016 package"
+	$(MAKE) -C ../ipasir/sat/picosat961 libipasirpicosat961.a
```
How can a solver be easily selected for use with ipasir?

nmanthey replied:
This requires a rebuild of CBMC. Solvers that implement the ipasir interface should be able to produce a library, which can be used during link time. The compilation of CBMC has to know about that file. Furthermore, each such rebuild has to make sure that the solver.a file is properly updated. Build instructions can be found in the top commit of this series.

tautschnig replied:
Could the build instructions be added to the "COMPILING" file, please?

peterschrammel commented:
IPASIR does not provide support for the preprocessors of Minisat and Glucose, which is a known issue. Hence, it can only be used with `--no-sat-preprocessing` in CBMC (or better set the option `sat-preprocessing` to false `#ifdef HAVE_IPASIR` so that the regression tests can be run normally). 
Still, I think that it is a valuable addition to easily benchmark more solvers.

nmanthey commented:
I rebased on top of origin/develop.

I fixed the pattern matching for the failing test I spotted so far - where the failure is usually due to the test matching an iteration number, which seems to depend on the SAT solver behaviour. I dropped these patterns in all tests in a new commit.

I also applied some suggestions from the reviewers, and modified the commands to build the solver for IPASIR support.

nmanthey commented:
I also fixed the wrong comment (copy paste error) that mentions minisat

nmanthey commented:
The problem reported in travis is not touched by my code path. As far as I can tell, I resolved all blockers that have been introduced with my patches! Please tell me how I can move forward!

https://travis-ci.org/diffblue/cbmc/builds/286241919?utm_source=github_status&utm_medium=notification

tautschnig commented:
@mgudemann Any chance you could review your change requests here?

tautschnig commented on regression/cbmc-cover/mcdc1/test.desc:
```diff
@@ -10,6 +10,6 @@ main.c
 ^\[main.coverage.6\] file main.c line 14 function main MC/DC independence condition `C && D && E && !A && !B.*: SATISFIED$
 ^\*\* .* of .* covered \(100.0%\)$
-^\*\* Used 10 iterations$
+^\*\* Used 
```
I believe we can safely remove that line. (And the same holds for all the other "Used" below.)

nmanthey replied:
I'll add another commit that drops all the lines that only report "Used". The other lines I cut still contain some useful information.

tautschnig commented on src/config.inc:
```diff
@@ -22,7 +22,8 @@ endif
 #MINISAT = ../../MiniSat-p_v1.14
-MINISAT2 = ../../minisat-2.2.1
+#MINISAT2 = ../../minisat-2.2.1
+#IPASIR = ../../ipasir
```
`LIBSOLVER` should be added here, alongside instructions on when to set it, and what to set it to.

nmanthey replied:
I am happy to document LIBSOLVER in that file, however, I would not force people to use a solver by default (yet), as all CI will fall apart. I'd make this a followup commit.

tautschnig commented on src/solvers/sat/satcheck_ipasir.cpp:
```diff
@@ -0,0 +1,315 @@
+
+ Purpose:
+
+\*******************************************************************/
```
Please use doxygen-style comment headers (or omit them).

nmanthey replied:
Will do.

tautschnig commented on src/solvers/sat/satcheck_ipasir.cpp:
```diff
@@ -0,0 +1,315 @@
+
+
+satcheck_ipasirt::satcheck_ipasirt()
+: solver(0)
```
Use `nullptr` instead of `0`

nmanthey replied:
will fix both.

tautschnig commented on src/solvers/sat/satcheck_ipasir.cpp:
```diff
@@ -0,0 +1,315 @@
+{
+  if(solver)
+    ipasir_release(solver);
+  solver=0;
```
`nullptr`

tautschnig commented on src/solvers/sat/satcheck_ipasir.cpp:
```diff
@@ -0,0 +1,315 @@
+      clause_counter << " clauses" << eom;
+  }
+
+    // use the internal representation, as ipasir does not support reporting the
```
Weird indentation

nmanthey replied:
Will fix.

nmanthey commented:
The current status addresses the comments raised by tautschnig, except forcing people to always choose a solver. I'd leave this decision to the project maintainers. I just wanted to add the opportunity to use IPASIR solvers.

tautschnig commented:
@mgudemann @peterschrammel @kroening can all of you please take a look at this?

kroening commented:
This is ready to go.

The only oddity is that for all other solvers, the solver gets linked into the solvers.a file. For ipasir, it is linked when the final executable is built. It might be slightly nicer to have that done uniformly.

romainbrenguier commented on src/solvers/sat/satcheck_ipasir.cpp:
```diff
@@ -0,0 +1,200 @@
+/*******************************************************************\
+
+Module:
```
Can you name this module?

romainbrenguier commented on src/solvers/sat/satcheck_ipasir.cpp:
```diff
@@ -0,0 +1,200 @@
+#include <ipasir.h>
+}
+
+/* solver interface:
```
I'm not sure this should be documented here.
Maybe add a link to the full documentation if it exists.
Otherwise it can stay.

romainbrenguier commented on src/solvers/sat/satcheck_ipasir.cpp:
```diff
@@ -0,0 +1,200 @@
+
+*/
+
+tvt satcheck_ipasirt::l_get(literalt a) const
```
Can you document what this functions does (and the one below as well).

romainbrenguier commented on src/solvers/sat/satcheck_ipasir.cpp:
```diff
@@ -0,0 +1,200 @@
+             "reject out of bound variables");
+  }
+
+
```
remove extra line

romainbrenguier replied:
or alternatively run clang-format-3.8 on the PR

romainbrenguier commented on src/solvers/sat/satcheck_ipasir.cpp:
```diff
@@ -0,0 +1,200 @@
+  {
+    if(!it->is_false())
+    {
+      // add literal wit correct sign
```
*with

romainbrenguier commented on src/solvers/sat/satcheck_ipasir.cpp:
```diff
@@ -0,0 +1,200 @@
+
+    forall_literals(it, assumptions)
+      if(it->is_false())
+        has_false=true;
```
I think ideally this should use `std::find` as it is not really using the iterator, but I'm not going to insist on that.

romainbrenguier commented on src/solvers/sat/satcheck_ipasir.cpp:
```diff
@@ -0,0 +1,200 @@
+        if(!it->is_false())
+          ipasir_assume(solver, it->sign()?-it->var_no():it->var_no());
+
+      if(10==ipasir_solve(solver))
```
what is the meaning of this number?

romainbrenguier commented on src/solvers/sat/satcheck_ipasir.cpp:
```diff
@@ -0,0 +1,200 @@
+
+bool satcheck_ipasirt::is_in_conflict(literalt a) const
+{
+  int v=a.var_no();
```
I think this is unsigned

romainbrenguier replied:
and const

romainbrenguier commented on src/solvers/sat/satcheck_ipasir.cpp:
```diff
@@ -0,0 +1,200 @@
+
+  if(ipasir_failed(solver, v))
+      return true;
+  return false;
```
what about `return ipasir_failed(solver, a.var_no());` ?

romainbrenguier commented on src/solvers/sat/satcheck_ipasir.cpp:
```diff
@@ -0,0 +1,200 @@
+  forall_literals(it, bv)
+    if(it->is_true())
+    {
+      assumptions.clear();
```
It seems assumptions can be cleared several times, I suggest using `std::find_if` instead.

nmanthey replied:
I'll replace it with a find_if.
However, the next line is a return, so that clear() is only called once.

tautschnig commented:
@reuk Just in case you have a moment to spare: I just looked at the clang-format suggestions in https://travis-ci.org/diffblue/cbmc/jobs/292240624 and it seems there is some missing indentation? Or is that just travis not getting quite right the output of leading space?

reuk commented:
> it seems there is some missing indentation? Or is that just travis not getting quite right the output of leading space?

I see this too, but if I run `git diff --color > patch.diff; cat patch.diff` locally the output has the correct number of spaces. It seems that Travis is collapsing one-or-more consecutive spaces into a single space.

tautschnig commented:
> It seems that Travis is collapsing one-or-more consecutive spaces into a single space.

Thank you very much for confirming! (Also I should have looked at the [raw log](https://s3.amazonaws.com/archive.travis-ci.org/jobs/292240624/log.txt?X-Amz-Expires=30&X-Amz-Date=20171025T085059Z&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAJRYRXRSVGNKPKO5A/20171025/us-east-1/s3/aws4_request&X-Amz-SignedHeaders=host&X-Amz-Signature=8caf762819d4a7fd5652e24c75f11e72f02a89c434101744b20c1721ee7c8661) yet that's pretty messy as well.)

reuk commented:
Maybe the solution is to disable colourising the diff - then at least the raw output will look right

nmanthey commented:
I applied the comments

tautschnig commented on src/solvers/sat/satcheck_ipasir.cpp:
```diff
@@ -0,0 +1,194 @@
+  else
+  {
+    // if assumptions contains false, we need this to be UNSAT
+    bvt::const_iterator it = std::find_if(assumptions.begin(), assumptions.end(), is_false);
```
Please line-wrap:
```
src/solvers/sat/satcheck_ipasir.cpp:114:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
```

nmanthey replied:
yes, spotted both of them locally. will install the linter githook before the next contribution ...

tautschnig commented on src/solvers/sat/satcheck_ipasir.cpp:
```diff
@@ -0,0 +1,194 @@
+      {
+        messaget::status() <<
+          "SAT checker: solving returned without solution" << eom;
+        throw "Solving inside IPASIR SAT solver has been interrupted";
```
Lowercase:
```
src/solvers/sat/satcheck_ipasir.cpp:146:  First character of throw error message should be lower case  [readability/throw] [4]
```

tautschnig commented:
> Maybe the solution is to disable colourising the diff - then at least the raw output will look right

True, but at the same time human readability would be impacted. I think we should stick with the current approach, unless @forejtv can work something similar to the Cpp Linter CI job, which would yield the raw diff in some easy-to-download form.

kroening commented on src/solvers/sat/satcheck_ipasir.cpp:
```diff
@@ -0,0 +1,194 @@
+    {
+      // add literal with correct sign
+      ipasir_add(solver, it->sign()?-it->var_no():it->var_no());
+    }
```
Another little tweak: the above can simply be it->dimacs()

kroening commented on src/solvers/sat/satcheck_ipasir.cpp:
```diff
@@ -0,0 +1,194 @@
+      forall_literals(it, assumptions)
+        if(!it->is_false())
+          ipasir_assume(solver, it->sign()?-it->var_no():it->var_no());
+
```
Use it->dimacs() as well.

kroening commented on src/solvers/sat/satcheck_ipasir.cpp:
```diff
@@ -0,0 +1,194 @@
+
+      // solve the formula, and handle the return code (10=SAT, 20=UNSAT)
+      int solver_state==ipasir_solve(solver);
+      if(10==solver_state)
```
I am surprised that the == parses?

nmanthey replied:
That slipped through copy and paste, and the compiler was happy ... if fixed it and tested another round with the LIBSOLVER build locally.

kroening commented on COMPILING.md:
````diff
@@ -46,6 +46,19 @@ We assume that you have a Debian/Ubuntu or Red Hat-like distribution.
+   Note: the LIBSOLVER variable could be pointed towards other solvers
+   ```
+   make -C src IPASIR=../../ipasir LIBSOLVER=$(pwd)/ipasir/libipasir.a
+   ```
````
We've got the names for the libraries of the other solvers in the config; why not for IPASIR as well?

Then it would be enough to say
make -C src IPASIR=../../ipasir

nmanthey replied:
Because typically (except you always want to use picosat) you want to push people to select their most favourite IPASIR solver explicitly at compile time.

kroening replied:
ok!

tautschnig commented:
@mgudemann @romainbrenguier Could you please review/approve the changes or add further comments?

mgudemann commented on COMPILING.md:
````diff
@@ -46,6 +46,19 @@ We assume that you have a Debian/Ubuntu or Red Hat-like distribution.
+
+   Get an IPASIR package and build picosat by default
+   ```
+   make -C ipasir-build
````
should be `make -C ipasir` which is the folder after doing `make -C src ipasir-download`

mgudemann replied:
`genipamax` hard-codes `g++-4.7`, unsure how to fix, maybe use just `g++` in the `C++11` file

File: genipamax/README
7:26:The makefiles expect g++-4.7, if you want to use a different compiler

```
NOTE: A C++ compiler supporting the std=c++0x flag is required.
The makefiles expect g++-4.7, if you want to use a different compiler
update the 'C++11' file.
```

mgudemann replied:
ok, but then again, this is only needed if one explicitely wants those solvers

nmanthey replied:
I agree. This is an external package. I will raise this for that package, so that it eventually gets resolved there.

mgudemann commented on src/solvers/sat/satcheck_ipasir.cpp:
```diff
@@ -0,0 +1,195 @@
+#include <inttypes.h>
+#endif
+
+#include <cassert>
```
I needed `<algorithm>` here, too, for `find_if`

tautschnig commented:
Assigning to @nmanthey for a final set of changes following @mgudemann's testing/comments.
